### ISSUE TITLE
Inflate: More diverse benchmarking corpus

### ISSUE THREAD

**Issue by Shnatsel** [etemesi254/zune-image#47]
#46 showed that zune-inflate is not (yet?) a win across the board - there was a situation when it wasn't better than even `miniz_oxide`!

A more diverse benchmarking corpus is needed to track performance more accurately, and avoid such situations in the future.

Ideas for what could be useful as benchmarks:
 - Compressed JSON, e.g. from `rustdoc`
 - Data stream from a PNG with a screenshot (to stress RLE on flat surfaces)
 - Data stream from a PNG with artwork (to stress complex patterns)
 - Something poorly compressible, like JPEG, that someone decided to compress anyway

**Comment by johannesvollmer** [comment]
Some infos about exr files, as requested:

### General Layout

EXR Images are split up into individual scan lines, or into blocks of arbitrary size (for example 64x64 pixel blocks, with smaller blocks at the edges). There are two variations how those blocks are compressed using zlib. Either line by line, or in batches, by batching 16 lines together. The width of the image is still unconstrained, so there is no hard limit. Each block is compressed and decompressed individually, to allow processing multiple blocks in parallel.

The contents of the streams are just preprocessed numbers, without any kind of meta data. The numbers are either `f32`, `f16`, or `u32`. They always occur batched together, for example 64 floats, followed by 64 integers, followed by 64 half floats. 

When compressing, the numbers are preprocessed with the following strategy:
1. bytes are shuffled to convert the numbers to little endian
2. bytes are separated such that every second byte is pushed to the second half of the byte slice (the goals is to increase the predictability of the byte sequence, by grouping together the most-significant digits and the least-significant digits)
3. instead of storing bytes, the differences between adjacent bytes is stored. This is done to increase the predictability of solid-colored areas, that have the same value over and over in one row

Have a look at this source file, it's only 40 lines:
[exrs/src/compression/zip.rs](https://github.com/johannesvollmer/exrs/blob/master/src/compression/zip.rs)

### Example Data

I prepared some [crow image zlib byte dumps (zipped google drive)](https://drive.google.com/file/d/1wlFkPvdMe01uQbz5_9KN83j8IL03_LQC/view?usp=sharing)

Taken from two images in the exrs repository:
1. `tests/images/valid/custom/crowskull/crow_zips.exr` (f32 values only, line by line)
2. `tests/images/valid/custom/crowskull/crow_zip_half.exr` (f16 values only, 16 lines batched)

Disclaimer: Didn't test whether this actually worked, pretty sure it should work. Here's the code I used, and you can use too, to extract the compressed byte streams from any suitable exr file:

```rust
use std::io::{BufReader};
use std::fs::File;
use exr::block::chunk::CompressedBlock;
use std::path::Path;
use exr::prelude::*;

/// Extracts the compressed byte blocks from an image.
/// They could be decompressed using the appropriate
fn main() {
    let out_folder_dir = Path::new("compressed_chunks/");
    let _ = std::fs::create_dir_all(out_folder_dir);

    let input_file = Path::new("tests/images/valid/custom/crowskull/crow_zip_half.exr");
    let file = BufReader::new(File::open(input_file).unwrap());

    // start reading the file, extracting the meta data of the image
    let image_reader = exr::block::read(file, true).unwrap();

    // print some debug info
    for header in &image_reader.meta_data().headers {
        match header.compression {
            Compression::ZIP1 => println!("image contains line-by-line zip compression"),
            Compression::ZIP16 => println!("image contains zip compression of 16 lines at once"),
            _ => panic!("image contains non-zip contents"),
        }

        println!("the image pixels are split up by: {:#?}", header.blocks);
    }

    // create a reader that loads all chunks from the file
    // note: chunks are the zip compressed pixel sections from the image
    let chunks = image_reader.all_chunks(false).unwrap();
    let chunk_count = chunks.len();

    // load the chunks without decompressing
    for (index, chunk) in chunks.enumerate() {
        let chunk = bytes_only_from_chunk(chunk.unwrap().compressed_block);

        let path = out_folder_dir.join(
            input_file.file_name().unwrap().to_str().unwrap().to_owned()
                + "_chunk_" + index.to_string().as_str() + ".bin"
        );

        println!("chunk #{} has {:?} bytes", index, chunk.len());
        println!("writing chunk {} to {:?}", index, &path);
        std::fs::write(&path, &chunk).unwrap();
    }

    println!("\nunpacked {} compressed chunks", chunk_count);
}

fn bytes_only_from_chunk(block: CompressedBlock) -> Vec<u8> {
    match block {
        CompressedBlock::ScanLine(block) => block.compressed_pixels,
        CompressedBlock::Tile(block) => block.compressed_pixels,
        CompressedBlock::DeepScanLine(block) => block.compressed_sample_data,
        CompressedBlock::DeepTile(block) => block.compressed_sample_data
    }
}
```

**Comment by etemesi254** [comment]
@Shnatsel , another issue might be that scanline decompression may be too small so we spend fewer times in the hot loop, eg some files in  the zip are 587 bytes.

Since I believe exr creates zlib file per scanline to allow for parallel scanline decompression, this means other factors may become dominant, i.e  table building time, we have a more complex table building operation than miniz-oxide and that the short runtime in the optimized loop doesn't account for most things.

Because copy_rep_matches is still appearing in samply and it's only used for the slower generic loop.

**Comment by Shnatsel** [comment]
@etemesi254 could you publish the latest revision of `zune-inflate` to crates.io? 

It represents a small but measurable improvement for loading images in EXR - decoding time down from 121ms to 113ms. I'd like to use it in the `exr` crate now that we have fuzzing on CI and we know it won't regress again.

**Comment by Shnatsel** [comment]
By the way, OpenEXR also has a compression mode with Zlib in batches of 16 scanlines instead of just one scanline. I suspect the benchmark is only for 1 scanline currently, and the one with batches of 16 will see a bigger improvement.

**Comment by etemesi254** [comment]
Done

**Comment by etemesi254** [comment]
I will dedicate time to improve speeds though

This is not the last from my bit io madness

**Comment by etemesi254** [comment]
> Compressed JSON, e.g. from rustdoc

Didn't know of a place I can get large json, so I replaced it with tokio/rs source code in gz format.
(tests/gz/tokio.tar.gz)

>Data stream from a PNG with artwork (to stress complex patterns)

Done (tests/zlib/png_artwork.zlib)

> Something poorly compressible, like JPEG, that someone decided to compress anyway

Done (tests/gzip/image_compressor/)

Will re-run whole benchmark and update site.

Tho the gist of it is that libdeflate got a lot of things right.

**Comment by Shnatsel** [comment]
To get a big JSON, you can run `cargo rustdoc -- -Z unstable-options --output-format json` on a crate with a large API surface.

For example, this is from `image` - 3.8 MB uncompressed, 420KB compressed. [image.json.gz](https://github.com/etemesi254/zune-image/files/10334850/image.json.gz)

**Comment by etemesi254** [comment]
Done, you can view speeds at https://etemesi254.github.io/assets/criterion/report/index.html

The gist is that.
1. We're bad at json, my hunch is that json is repetitive and hence we have the RLE decoders which are bounds checked, I'm gonna investigate this though.

2. We're mid at xml/html, better than zlib-ng

3. We're not so bad on png zlibs

4. We are not so bad on source codes gzips.


But the criterion benchmarks provide more detauls
